cdap-data-fabric/src/main/java/co/cask/cdap/data/format/AvroRecordFormat.java

```diff
@@ -0,0 +1,71 @@
+/*
+  * Copyright © 2015 Cask Data, Inc.
+  *
+  * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+  * use this file except in compliance with the License. You may obtain a copy of
+  * the License at
+  *
+  * http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+  * License for the specific language governing permissions and limitations under
+  * the License.
+  */
+
+package co.cask.cdap.data.format;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.data.format.UnexpectedFormatException;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.data.schema.UnsupportedTypeException;
+import org.apache.avro.SchemaParseException;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.DecoderFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+/**
+ * Stream record format that interprets the body as avro encoded binary data.
+ */
+public class AvroRecordFormat extends ByteBufferRecordFormat<GenericRecord> {
+  private org.apache.avro.Schema avroSchema;
+  private DecoderFactory decoderFactory = DecoderFactory.get();
+  private DatumReader<GenericRecord> datumReader;
+
+  @Override
+  public GenericRecord read(ByteBuffer input) {
+    try {
+      return datumReader.read(null, decoderFactory.binaryDecoder(Bytes.toBytes(input), null));
```

Comment:
Is it worth figuring out a way in the future to re-use the GenericRecord and BinaryDecoder? I don't know how expensive these are to construct, but for map reduce usage this could be a fair amount of object churn.  If it seems worthwhile, it would be good to add a comment or open a JIRA just to track it.